SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -2428,7 +2428,32 @@ void USpatialReceiver::PeriodicallyProcessIncomingRPCs()
 
 bool USpatialReceiver::NeedToLoadClass(const FString& ClassPath)
 {
-	return FindObject<UClass>(nullptr, *ClassPath, false) == nullptr;
+	UObject* ClassObject = FindObject<UClass>(nullptr, *ClassPath, false);
+	if (!ClassObject)
+	{
+		return true;
+	}
+
+	FString PackagePath = GetPackagePath(ClassPath);
+	FName PackagePathName = *PackagePath;
+
+	// UNR-3320 The following test checks if the package is currently being processed in the async loading thread.
+	// Without it, we could be using an object loaded in memory, but not completely ready to be used.
+	// Looking through PackageMapClient's code, which handles asset async loading in Native unreal, checking
+	// UPackage::IsFullyLoaded, or UObject::HasAnyInternalFlag(EInternalObjectFlag::AsyncLoading) should tell us if it is the case.
+	// In practice, these tests are not enough to prevent using objects too early (symptom is RF_NeedPostLoad being set, and crash when using them later).
+	// GetAsyncLoadPercentage will actually look through the async loading thread's UAsyncPackage maps to see if there are any entries.
+	// TODO : UNR-XXXX This looks like an expensive check, but it does the job. We should investigate further 
```

Comment:
Please create this ticket

Reply:
https://improbableio.atlassian.net/browse/UNR-3374